<a href="https://colab.research.google.com/github/bakut/packt-data-science-workshop/blob/master/Building_a_Model_for_classifying_Animal_Type_and_assessing_its_performance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Building a Model for Classifying Animal Type and Assessing Its Performance**


In [0]:
#importing pands as pd
import pandas as pd

In [0]:
#creating a variable to contain the URL of the dataset
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter04/Dataset/openml_phpZNNasq.csv'

In [32]:
#load dataset into a dataframe
df = pd.read_csv(file_url)
df.head()

animal   hair  feathers   eggs  ...   tail  domestic  catsize    type
0  aardvark   True     False  False  ...  False     False     True  mammal
1  antelope   True     False  False  ...   True     False     True  mammal
2      bass  False     False   True  ...   True     False    False    fish
3      bear   True     False  False  ...  False     False     True  mammal
4      boar   True     False  False  ...   True     False     True  mammal

[5 rows x 18 columns]

using the type column as our target variable. remove the animal column from the dataframe and only use the remaing columns as features.

In [0]:
#remove the animal column .drop() method from pandas
df.drop(columns= 'animal', inplace=True)

In [0]:
#Extract the type column using .pop() method from pandas
y = df.pop('type')

In [22]:
df.head()

hair  feathers   eggs   milk  ...  legs   tail  domestic  catsize
0   True     False  False   True  ...     4  False     False     True
1   True     False  False   True  ...     4   True     False     True
2  False     False   True  False  ...     0   True     False    False
3   True     False  False   True  ...     4  False     False     True
4   True     False  False   True  ...     4   True     False     True

[5 rows x 16 columns]

In [0]:
#import the train_test_split function from sklearn.model_selection
from sklearn.model_selection import train_test_split

In [0]:
#split the dataset into training and testing set
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.4, random_state=188)

In [0]:
#import RandomForestClassifier from sklearn.ensemble
from sklearn.ensemble import RandomForestClassifier

In [0]:
#instantiate the RandomForestClassifier object with random_state equal to 42
rf_model = RandomForestClassifier(random_state=42)

In [27]:
#Fit RandomForestClassifier with the training set
rf_model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [28]:
#predict the outcome of the training set with the .fit()
train_preds = rf_model.predict(X_train)
train_preds

array(['mammal', 'mammal', 'mammal', 'fish', 'mammal', 'insect', 'fish',
       'bird', 'mammal', 'mammal', 'fish', 'bird', 'reptile', 'bird',
       'fish', 'mammal', 'mammal', 'bird', 'bird', 'mammal', 'bird',
       'bird', 'mammal', 'invertebrate', 'reptile', 'invertebrate',
       'fish', 'bird', 'mammal', 'mammal', 'amphibian', 'mammal',
       'invertebrate', 'mammal', 'mammal', 'insect', 'mammal', 'fish',
       'invertebrate', 'mammal', 'invertebrate', 'invertebrate', 'insect',
       'amphibian', 'mammal', 'reptile', 'amphibian', 'invertebrate',
       'mammal', 'fish', 'bird', 'mammal', 'mammal', 'bird', 'mammal',
       'mammal', 'fish', 'mammal', 'bird', 'fish'], dtype=object)

In [0]:
#import the accuracy_score function from sklearn.metrics
from sklearn.metrics import accuracy_score

In [0]:
#calculate the accuracy score
train_acc = accuracy_score(y_train, train_preds)

In [31]:
print(train_acc)

1.0


Our model achieved an accuracy of 1 on the training set, which means it perfectly predicted the target variable on all of those observations. Let's check the performance on the testing set.

In [0]:
#predict the outcome of the testing set with the .fit() method
test_preds = rf_model.predict(X_test)

In [34]:
#calculate the accuracy score on the testing set and save the result variable called test_acc
test_acc = accuracy_score(y_test, test_preds)
print(test_acc)

0.9024390243902439


we trained a RandomForest to predict the type of animals based on their key attributes. Our model achieved a perfect accuracy score of 1 on the training set but only 0.90 on the testing set. This means our model is overfitting and is not general enough. The ideal situation would be for the model to achieve a very similar, high-accuracy score on both the training and testing sets.